// cloudflare-allowme/lib/settings.js
"use strict";
if (!process.env.CF_ALLOWME_TOKEN) {
    require("dotenv").config();
}
const settings = {
    cloudflare: {
        token: process.env.ALLOWME_CF_TOKEN || "",
        zone: process.env.ALLOWME_CF_ZONE || "",
        accountId: process.env.ALLOWME_CF_ACCOUNTID || "",
        zoneId: null,
        firewallId: null,
        listId: process.env.ALLOWME_CF_LISTID || "",
        listName: null
    },
    server: {
        port: process.env.ALLOWME_SERVER_PORT || "",
        secret: process.env.ALLOWME_SERVER_SECRET || "",
        user: process.env.ALLOWME_SERVER_USER || "",
        prompt: process.env.ALLOWME_SERVER_PROMPT || "",
        trustProxy: process.env.ALLOWME_SERVER_TRUSTPROXY || "",
        home: process.env.ALLOWME_SERVER_HOME || ""
    },
    ip: {
        maxAge: process.env.ALLOWME_IP_MAXAGE || "",
        blockInterval: process.env.ALLOWME_IP_BLOCKINTERVAL || "",
        denyCount: process.env.ALLOWME_IP_DENYCOUNT || ""
    },
    log: {
        level: process.env.ALLOWME_LOG_LEVEL || ""
    }
};
module.exports = settings;
